examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,892 @@
+
+```rust
+#[derive(Debug, Clone, FromSqlRow, AsExpression, PartialEq, Eq)]
+#[diesel(sql_type = PgProtocolType)]
+pub enum ProtocolType {
+    UnknownProtocol,
+    GRPC,
+    HTTP,
+    UDP,
+}
+```
+
+It’s worth pointing out that you are dealing with three types in total:
+
+1) protocol_type: The Postgres Enum type
+2) PGProtocolType: The Wrapper struct that refers to the Postgres Enum type
+3) ProtocolType: The Rust Enum that refers to the wrapper struct PGProtocolType
+
+Mixing any of those three types will result in a complicated Diesel trait error. However, these error messages are just
+a convoluted way to say that the database type mismatches the Rust type. When you encounter a consulted trait error
+message, make sure to check:
+
+1) Do I have a wrapper struct?
+2) Does my a wrapper struct derives SqlType?
+3) Does my wrapper type has both, sql_type and postgres_type declared?
+4) Are sql_type and postgres_type both referring to the correct Postgres type?
+5) Does my Rust type refers to the wrapper struct type?
+
+If all those checks pass and you still see errors, it’s most likely a serialization error.
+
+To serialize and deserialize a custom Enum, you write a custom toSql and fromSql implementation. Luckily, this is
+straightforward.
+
+```rust
+
+impl ToSql<PgProtocolType, Pg> for ProtocolType {
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> serialize::Result {
+        match *self {
+            ProtocolType::UnknownProtocol => out.write_all(b"UnknownProtocol")?,
+            ProtocolType::GRPC => out.write_all(b"GRPC")?,
+            ProtocolType::HTTP => out.write_all(b"HTTP")?,
+            ProtocolType::UDP => out.write_all(b"UDP")?,
+        }
+        Ok(IsNull::No)
+    }
+}
+
+impl FromSql<PgProtocolType, Pg> for ProtocolType {
+    fn from_sql(bytes: PgValue<'_>) -> deserialize::Result<Self> {
+        match bytes.as_bytes() {
+            b"UnknownProtocol" => Ok(ProtocolType::UnknownProtocol),
+            b"GRPC" => Ok(ProtocolType::GRPC),
+            b"HTTP" => Ok(ProtocolType::HTTP),
+            b"UDP" => Ok(ProtocolType::UDP),
+            _ => Ok(ProtocolType::UnknownProtocol),
+        }
+    }
+}
+```
+
+In toSql, it is important to end the implementation with Ok(IsNull::No). In the from_sql, it is important to add a catch
+all case that returns an UnknownProtocol instance. In practice, the catch all rarely ever gets triggered, but in those
+few corner cases when it does, it keeps the application running.
+
+### Rust Struct Endpoint
+
+```rust 
+#[derive(Debug, Clone, FromSqlRow, AsExpression, PartialEq, Eq)]
+#[diesel(sql_type=crate::schema::smdb::sql_types::ServiceEndpoint)]
+pub struct Endpoint {
+    pub name: String,
+    pub version: i32,
+    pub base_uri: String,
+    pub port: i32,
+    pub protocol: ProtocolType,
+}
+```
+
+It is worth mentioning is that the sql_type refers to the wrapper struct generated by Diesel and stored in the schema.rs
```

Comment:
> wrapper struct

We call that type sql side type in our documentation. (At least we try to be consistent)

Reply:
Good to know